app/embedding.py

```diff
@@ -0,0 +1,66 @@
+import os
+import torch
+from torch.cuda.amp import autocast
+import pickle
+import numpy as np
+from transformers import AutoModel
+from sentence_transformers import SentenceTransformer
+from processor import Processor
+
+FILE_PATH = os.path.dirname(os.path.abspath(__file__))
+os.environ["PYTORCH_CUDA_ALLOC_CONF"] = 'max_split_size_mb:256'
+
+class Embedding:
+    def __init__(self, model_name):
+        self.model = AutoModel.from_pretrained(model_name, trust_remote_code=True,
+                                               torch_dtype=torch.bfloat16)
+
+    def main(self):
+
+        # Chunk the text
+        chunks = Processor().chunk_text(self.text, spacy_model='de_core_news_lg')
+
+        # Generate embeddings for each chunk
+        embeddings = self.create_embeddings(chunks)
+
+        # Store chunks and embeddings
+        self.save_chunks(chunks)
+        self.save_embeddings(embeddings)
+
+
+    def create_embeddings(self, chunks, batch_size=8):
+        """Generate embeddings for each chunk using Sentence-BERT."""
+
+        # Catch edge case if there are no chunks
+        if not chunks:
+            return []
+
+        embeddings = []
+        with torch.no_grad():  # Disable gradient calculations
+            for i in range(0, len(chunks), batch_size):
+                batch_chunks = chunks[i:i + batch_size]
+                with autocast():  # Enable mixed precision
+                    batch_embeddings = self.model.encode(batch_chunks)  # Generate embeddings for the batch
+                embeddings.extend(batch_embeddings)  # Append batch embeddings to the main list
+
+                # Clear CUDA cache after each batch to free memory
+                torch.cuda.empty_cache()
+
+        return embeddings
+
+    @staticmethod
+    def save(embeddings):
+        np.save(os.path.join(FILE_PATH, '../data/embeddings.npy'), embeddings)
+
+    @staticmethod
+    def load(path= os.path.join(FILE_PATH, '../data/embeddings.npy')):
+        embeddings = np.load(path)
+        return embeddings
+
+if __name__ == "__main__":
+    # model_name='jinaai/jina-embeddings-v2-small-en'
+    model_name = 'jinaai/jina-embeddings-v2-base-de'
+    embedding = Embedding(model_name=model_name)
+    embedding.main()
```

Comment:
This whole code would be unnecessary I assume, since it is only required for the method Embeddings.create_new_embeddings() highlighted below.

Reply:
Moving the `update-es-index` step from `workflow.py` to the setup_instructions sounds good to me!

Regarding the create_new_embeddings() method, it is only required once to create the embeddings for the `data/legal-basis.txt` using:

```bash
python workflow.py create-embeddings
```

It takes the `data/legal-basis.txt` as input, chunks it into 4kb parts, and computes the embeddings using `jinaai/jina-embeddings-v2-base-de`. This is also an operation I had to do only once, because the result is stored inside `data/embeddings.npy` and is simply loaded for future vector-search.

I took the same logic as for the `update-es-index` step, we only create it once. But here the difference is that the embedding model is quite huge, and it took 1) very long and 2) lots of resources to compute. Therefore I stored the results in a numpy array to later on load it easily. 

If the `data/legal-basis.txt` text changes in the future, we would need to re-run the `create-embeddings` workflow to regenerate the embeddings. 

Let me know if you have suggestions on how to do it differently, thanks again for pointing this out! 💯

Reply:
Thanks for the explanation! I added this quickly to the README, so we have this information easily accessible and do not lose it. To me, you can approve and merge this branch now as long as you don't see any improvements for the latest commit I did.

Reply:
Great, I made one minor modification in the README, I will go ahead and approve the MR! Thanks 🚀